File: src/core/run/shop.ts
// 상점 (7.2) — 골드 구매. 강화권은 상점에서만(4.6). 진열 생성 + 구매 + 나가기.
import type { PartyMemberState } from "../types.ts";
import type { RunState, ShopOffer } from "./types.ts";
import { CHARACTERS } from "../../data/characters.ts";
import { SKILLS } from "../../data/skills.ts";
import { unlockedTier, ownsUpgradeLine } from "./rewards.ts";
import { genItemOffers } from "./items.ts";
import { healParty, upgradeOwned, learnOwned, completeNode } from "./helpers.ts";

export function generateShop(run: RunState): ShopOffer[] {
  let k = 0;
  const mk = () => `shop${run.visited.length}_${k++}`;
  const pool: ShopOffer[] = [];
  const tierOk = (m: PartyMemberState, sid: string) => !run.useMastery || (SKILLS[sid]?.tier ?? 1) <= unlockedTier(m.masteryLevel); // 숙련도 tier 게이팅(4.4)
  for (const m of run.party.filter((p) => p.hp > 0)) {
    const c = CHARACTERS[m.charId];
    for (const sid of m.ownedSkillIds) {
      const sk = SKILLS[sid];
      const to = sk?.nextTierId ? SKILLS[sk.nextTierId] : undefined;
      if (sk && to && tierOk(m, sk.nextTierId!)) pool.push({ id: mk(), kind: "upgrade", cost: 25, charId: m.charId, fromSkillId: sid, toSkillId: sk.nextTierId!, label: `강화권: ${c.name} 「${sk.name}」→「${to.name}」` });
    }
    for (const sid of c.skillIds) {
      if (!ownsUpgradeLine(m.ownedSkillIds, sid) && !SKILLS[sid]?.exclusiveTo && tierOk(m, sid)) pool.push({ id: mk(), kind: "learn", cost: 20, charId: m.charId, skillId: sid, label: `스킬: ${c.name} 「${SKILLS[sid].name}」(범용)` });
    }
  }
  const picked: ShopOffer[] = [];
  while (picked.length < 3 && pool.length) picked.push(pool.splice(run.rng.int(0, pool.length - 1), 1)[0]);
  picked.push(...genItemOffers(run, mk, 2)); // 장착 아이템 진열 (4.3)
  picked.push({ id: mk(), kind: "heal", cost: 15, pct: 0.5, label: "치료: 파티 50% 회복" });
  return picked;
}

export function buyShopOffer(run: RunState, offerId: string): void {
  if (run.phase !== "shop" || !run.shop) return;
  const o = run.shop.find((x) => x.id === offerId);
  if (!o || run.gold < o.cost) return;
  run.gold -= o.cost;
  if (o.kind === "upgrade") { const m = run.party.find((p) => p.charId === o.charId); if (m) upgradeOwned(m, o.fromSkillId, o.toSkillId); }
  else if (o.kind === "learn") { const m = run.party.find((p) => p.charId === o.charId); if (m) learnOwned(m, o.skillId); }
  else if (o.kind === "buyItem") run.inventory.push(o.itemId); // 인벤토리에 추가 → 시트에서 장착
  else if (o.kind === "heal") healParty(run, o.pct);
  run.shop = run.shop.filter((x) => x.id !== offerId); // 구매한 항목 제거(재구매 방지)
  run.log.push(`구매: ${o.label} (−${o.cost}G)`);
}

export function leaveShop(run: RunState): void {
  if (run.phase !== "shop") return;
  run.shop = null;
  completeNode(run, run.activeNodeId!);
}

File: engine/spr-core/src/run/shop.rs
//! 상점 (TS `core/run/shop.ts`, 7.2) — 진열 생성(저작+절차) + 구매 + 나가기. RNG 추첨 결정론.
use super::data::RunData;
use super::helpers::{complete_node, heal_party, learn_owned, upgrade_owned};
use super::items::gen_item_offers;
use super::layers::advance_core;
use super::rewards::{owns_upgrade_line, reward_gate_ok};
use super::types::{RunState, ShopOffer};
use spr_types::map::ShopOfferDef;
use spr_types::party::PartyMemberState;

fn tier_ok(run: &RunState, m: &PartyMemberState, sid: &str, d: &RunData) -> bool {
    d.skills.get(sid).map(|sk| reward_gate_ok(sk, run.use_mastery, m.mastery_level, m.class_tier)).unwrap_or(false)
}

fn materialize_offers(run: &RunState, defs: &[ShopOfferDef], mk: &mut dyn FnMut() -> String, d: &RunData) -> Vec<ShopOffer> {
    let mut out = Vec::new();
    for sd in defs {
        match sd {
            ShopOfferDef::BuyItem { item_id, cost } => {
                if let Some(it) = d.items.get(item_id) {
                    out.push(ShopOffer::BuyItem { id: mk(), cost: *cost, item_id: it.id.clone(), label: format!("「{}」", it.name) });
                }
            }
            ShopOfferDef::Heal { pct, cost } => out.push(ShopOffer::Heal { id: mk(), cost: *cost, pct: *pct, label: format!("치료: 파티 {}% 회복", pct) }),
            ShopOfferDef::Learn { char_id, skill_id, cost } => {
                if let (Some(c), Some(sk)) = (d.chars.get(char_id), d.skills.get(skill_id)) {
                    if run.party.iter().any(|p| &p.char_id == char_id && p.hp > 0) {
                        out.push(ShopOffer::Learn { id: mk(), cost: *cost, char_id: char_id.clone(), skill_id: skill_id.clone(), label: format!("스킬: {} 「{}」", c.name, sk.name) });
                    }
                }
            }
        }
    }
    out
}

fn generate_procedural(run: &mut RunState, mk: &mut dyn FnMut() -> String, d: &RunData) -> Vec<ShopOffer> {
    let living: Vec<PartyMemberState> = run.party.iter().filter(|m| m.hp > 0).cloned().collect();
    let mut pool: Vec<ShopOffer> = Vec::new();
    for m in &living {
        let c = &d.chars[&m.char_id];
        for sid in &m.owned_skill_ids {
            if let Some(sk) = d.skills.get(sid) {
                if let Some(next) = &sk.next_tier_id {
                    if let Some(to) = d.skills.get(next) {
                        if tier_ok(run, m, next, d) {
                            pool.push(ShopOffer::Upgrade { id: mk(), cost: 25, char_id: m.char_id.clone(), from_skill_id: sid.clone(), to_skill_id: next.clone(), label: format!("강화권: {} 「{}」→「{}」", c.name, sk.name, to.name) });
                        }
                    }
                }
            }
        }
        for sid in &c.skill_ids {
            let sk = match d.skills.get(sid) {
                Some(s) => s,
                None => continue,
            };
            if !owns_upgrade_line(&m.owned_skill_ids, sid, &d.skills) && sk.exclusive_to.is_none() && tier_ok(run, m, sid, d) {
                pool.push(ShopOffer::Learn { id: mk(), cost: 20, char_id: m.char_id.clone(), skill_id: sid.clone(), label: format!("스킬: {} 「{}」(범용)", c.name, sk.name) });
            }
        }
    }
    let mut picked: Vec<ShopOffer> = Vec::new();
    while picked.len() < 3 && !pool.is_empty() {
        let idx = run.rng.int(0, pool.len() as i64 - 1) as usize;
        picked.push(pool.remove(idx));
    }
    picked.extend(gen_item_offers(run, mk, 2, &d.items, &d.item_pool));
    picked.push(ShopOffer::Heal { id: mk(), cost: 15, pct: 50, label: "치료: 파티 50% 회복".to_string() });
    picked
}

/// 노드 상점 진열 — offers 있으면 저작(+keepGenerated 절차), 없으면 절차. TS generateShop.
pub fn generate_shop(run: &mut RunState, offers: Option<Vec<ShopOfferDef>>, keep_generated: bool, d: &RunData) -> Vec<ShopOffer> {
    let visited_len = run.visited.len();
    let mut k = 0i64;
    let mut mk = move || {
        let s = format!("shop{}_{}", visited_len, k);
        k += 1;
        s
    };
    if let Some(o) = &offers {
        if !o.is_empty() {
            let authored = materialize_offers(run, o, &mut mk, d);
            if keep_generated {
                let mut a = authored;
                a.extend(generate_procedural(run, &mut mk, d));
                return a;
            }
            return authored;
        }
    }
    generate_procedural(run, &mut mk, d)
}

/// 상점 구매. TS buyShopOffer.
pub fn buy_shop_offer(run: &mut RunState, offer_id: &str, d: &RunData) {
    if run.phase != "shop" {
        return;
    }
    let o = match &run.shop {
        Some(s) => match s.iter().find(|x| offer_id_of(x) == offer_id) {
            Some(o) => o.clone(),
            None => return,
        },
        None => return,
    };
    let cost = cost_of(&o);
    if run.gold < cost {
        return;
    }
    run.gold -= cost;
    match &o {
        ShopOffer::Upgrade { char_id, from_skill_id, to_skill_id, .. } => {
            if let Some(mi) = run.party.iter().position(|p| &p.char_id == char_id) {
                upgrade_owned(&mut run.party[mi], from_skill_id, to_skill_id);
            }
        }
        ShopOffer::Learn { char_id, skill_id, .. } => {
            if let Some(mi) = run.party.iter().position(|p| &p.char_id == char_id) {
                learn_owned(&mut run.party[mi], skill_id);
            }
        }
        ShopOffer::BuyItem { item_id, .. } => run.inventory.push(item_id.clone()),
        ShopOffer::Heal { pct, .. } => heal_party(run, *pct, false, d),
    }
    if let Some(s) = &mut run.shop {
        s.retain(|x| offer_id_of(x) != offer_id);
    }
    run.log.push(format!("구매: {} (−{}G)", label_of(&o), cost));
}

/// 상점 나가기 → 시퀀스 복귀(coreCursor) 또는 노드 완료. TS leaveShop.
pub fn leave_shop(run: &mut RunState, d: &RunData) {
    if run.phase != "shop" {
        return;
    }
    run.shop = None;
    if run.core_cursor.is_some() {
        advance_core(run, d);
    } else if let Some(aid) = run.active_node_id.clone() {
        complete_node(run, &aid, d);
    }
}

fn offer_id_of(o: &ShopOffer) -> &str {
    match o {
        ShopOffer::Upgrade { id, .. } | ShopOffer::Learn { id, .. } | ShopOffer::BuyItem { id, .. } | ShopOffer::Heal { id, .. } => id,
    }
}
fn cost_of(o: &ShopOffer) -> i64 {
    match o {
        ShopOffer::Upgrade { cost, .. } | ShopOffer::Learn { cost, .. } | ShopOffer::BuyItem { cost, .. } | ShopOffer::Heal { cost, .. } => *cost,
    }
}
fn label_of(o: &ShopOffer) -> &str {
    match o {
        ShopOffer::Upgrade { label, .. } | ShopOffer::Learn { label, .. } | ShopOffer::BuyItem { label, .. } | ShopOffer::Heal { label, .. } => label,
    }
}
